drivers/dac/max2201x/iio_max2201x.c

```diff
@@ -0,0 +1,824 @@
+		return max2201x_reg_update(desc->max2201x_desc, MAX2201X_AO_CMD,
+					   MAX2201X_LD_CNFG_MASK(i / 2), no_os_field_prep(MAX2201X_LD_CNFG_MASK(i / 2),
+							   i % 2));
+
+	return max2201x_reg_update(desc->max2201x_desc, MAX2201X_GEN_CNFG,
+				   MAX2201X_IIO_CNFG_MASK(priv, i / 2),
+				   no_os_field_prep(MAX2201X_IIO_CNFG_MASK(priv, i / 2), i % 2));
+}
+
+static int max2201x_iio_read_config_available(void *dev, char *buf,
+		uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv)
+{
+	return max2201x_iio_read_avail_length(max2201x_config_enable_avail, buf);
+}
+
+static int max2201x_iio_write_timeout(void *dev, char *buf, uint32_t len,
+				      const struct iio_ch_info *channel,
+				      intptr_t priv)
+{
+	struct max2201x_iio_desc *desc = dev;
+	uint32_t i, mask;
+
+	switch (priv) {
+	case MAX2201X_IIO_TIMEOUT:
+		i = (uint32_t)max2201x_iio_read_avail(max2201x_timeout_avail, buf);
+
+		return max2201x_reg_update(desc->max2201x_desc, MAX2201X_GEN_CNFG,
+					   MAX2201X_TIMEOUT_MASK, no_os_field_prep(MAX2201X_TIMEOUT_MASK, i));
+	case MAX2201X_IIO_TIMEOUT_EN:
+		mask = MAX2201X_TIMEOUT_EN_MASK;
+		break;
+	case MAX2201X_IIO_TIMEOUT_CFG:
+		mask = MAX2201X_TIMEOUT_CNFG_MASK;
+		break;
+	default:
+		return -EINVAL;
+	}
+	i = (uint32_t)max2201x_iio_read_avail(max2201x_enable_avail, buf);
+
+	return max2201x_reg_update(desc->max2201x_desc, MAX2201X_GEN_CNFG, mask,
+				   no_os_field_prep(mask, i));
+}
+
+static int max2201x_iio_read_timeout_available(void *dev, char *buf,
+		uint32_t len, const struct iio_ch_info *channel, intptr_t priv)
+{
+	return max2201x_iio_read_avail_length(max2201x_timeout_avail, buf);
+}
+
+static int max2201x_iio_read_interrupt(void *dev, char *buf, uint32_t len,
+				       const struct iio_ch_info *channel,
+				       intptr_t priv)
+{
+	struct max2201x_iio_desc *desc = dev;
+	int ret;
+	uint32_t reg_val;
+
+	ret = max2201x_reg_read(desc->max2201x_desc, MAX2201X_GEN_INT, &reg_val);
+	if (ret)
+		return ret;
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, (int32_t *)&reg_val);
+}
+
+static int max2201x_iio_write_int_en(void *dev, char *buf, uint32_t len,
+				     const struct iio_ch_info *channel,
+				     intptr_t priv)
+{
+	struct max2201x_iio_desc *desc = dev;
+	int32_t val;
+
+	iio_parse_value(buf, IIO_VAL_INT, &val, NULL);
+
+	if (val < 0)
+		return -EINVAL;
+
+	return max2201x_reg_write(desc->max2201x_desc, MAX2201X_GEN_INTEN,
+				  (uint32_t)val);
+}
+
+static int max2201x_iio_reg_read(struct max2201x_iio_desc *iio_desc,
+				 uint32_t reg, uint32_t *readval)
+{
+	return max2201x_reg_read(iio_desc->max2201x_desc, reg, readval);
+}
+
+static int max2201x_iio_reg_write(struct max2201x_iio_desc *iio_desc,
+				  uint32_t reg, uint32_t writeval)
+{
+	return max2201x_reg_write(iio_desc->max2201x_desc, reg, writeval);
+}
+
+int max2201x_iio_read_avail(char *array_avail, char *buf)
+{
+	int i;
+
+	for (i = 0; i < NO_OS_ARRAY_SIZE(array_avail); i++)
+		if (!strcmp(buf, array_avail[i]))
```

Comment:
Is this working properly?

Reply:
This is my bad here, indeed it shouldn't be working properly, my guess when I last tested it, it may have worked by coincidence and treated the result as a good one. Will test again with the hardware connected and make sure this function works properly in all the cases and the if condition doesn't check by accident.